File: incre-tests/fusion/algprog/page60.f
Inductive Tree = tip Int | bin {Tree, Tree};

tri = \op: Int -> Int. fix (
  \f: Tree -> Compress Tree. \t: Tree.
  match t with
    tip _ -> tip 0
  | bin {l, r} ->
    let step = fix (
      \g: Tree -> Tree. \ys: Tree.
      match ys with
        tip w -> tip (op w)
      | bin {l, r} -> bin {g l, g r}
      end
    ) in bin {step (f l), step (f r)}
  end
);

op = \x: Int. + 1 x;

max = \a: Int. \b: Int. if < a b then b else a;

maximum = fix (
  \f: Tree -> Int. \t: Tree.
  match t with
    tip w -> w
  | bin {l, r} -> max (f l) (f r)
  end
);

main = \t: Tree. maximum (tri op t);

File: ml-benchmark/fusion/algprog/page60.ml
type tree = Tip of int | Bin of tree * tree

let rec tri op =
  let rec f t =
    match t with
    | Tip _ -> Tip 0
    | Bin (l, r) ->
        let rec g ys =
          match ys with
          | Tip w -> Tip (op w)
          | Bin (l2, r2) -> Bin (g l2, g r2)
        in
        Bin (g (f l), g (f r))
  in
  f

let op x = x + 1

let max a b = if a < b then b else a

let rec maximum t =
  match t with
  | Tip w -> w
  | Bin (l, r) -> max (maximum l) (maximum r)

let program t = maximum (tri op t)